torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,234 @@
+                                    )
+                                artifacts.append(
+                                    TritonKernelArtifact(filename, payload)
+                                )
+                            counters["inductor"]["triton_bundler_save_kernel"] += 1
+                        except Exception:
+                            log.debug("failed to collect triton kernel", exc_info=True)
+                        if filename.endswith(".cubin"):
+                            # Each kernel has bunch of files like .cubin, .json, .ttir
+                            # Just append one of them without the extension
+                            kernel_names.append(Path(filename).stem)
+                    if artifacts:
+                        result.append(
+                            TritonKernelArtifacts(
+                                entry.kernel_hash,
+                                entry.device,
+                                artifacts,
+                            )
+                        )
+                cls._entries = None
+                return result, TritonBundlerMetadata(kernel_names)
+            return [], None
+
+    @staticmethod
+    def read_and_emit(
+        bundle: List[TritonKernelArtifacts],
+    ) -> Optional[TritonBundlerMetadata]:
+        """
+        This is the main function called when a cache read happens. This function
+        converts the bundled format back into individual files and writes them
+        to the filesystem.
+
+        NOTE: When we are writing to the filesystem, we assume exclusive access
+        to the target directory.
+        This means that if the target folder already exists and is non-empty,
+        we bail out.
+        Exclusive access means that no other process should be writing to
+        or reading from the target directory.
+        """
+        if not TritonBundler.is_enabled():
+            return None
+
+        with dynamo_timed(key="TritonBundler.read_and_emit", fwd_only=False):
+            kernel_names: List[str] = []
+
+            for artifacts in bundle:
+                basedir = triton_cache_dir(artifacts.device)
+                directory = os.path.join(basedir, artifacts.kernel_hash)
+
+                if os.path.exists(directory) and len(os.listdir(directory)) != 0:
+                    # If directory already exists, we bail out and leave
+                    # local disk to take care of caching
+                    log.debug(
+                        "Bailing out TritonBundler.read_and_emit, %s is non empty",
+                        directory,
+                    )
+                    continue
+
+                Path(directory).mkdir(parents=True, exist_ok=True)
+
+                # Random ID to avoid any collisions
+                rnd_id = str(uuid.uuid4())
+                tmp_dir = os.path.join(basedir, f"tmp.{rnd_id}")
+                os.makedirs(tmp_dir)
```

Comment:
nit: uuid to generate tempfile is not actually the recommended way. On POSIX, you can use mkstemp or similar to create a temporary file in a directory that is guaranteed not to conflict with any other file.

Reply:
But actually, because you have exclusive access, you don't even need a random ID, you can just write to a well known temp file name and because you assume no one else is mucking with this dir it's fine. You only need to guard against premature termination

Reply:
although you do have to delete stale tmp dir if there are any lying around :P If it's a well known name, it's easier to make sure that you cleanup the next time around.